lib/init.php

```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
 	);
 }
 add_action( 'admin_menu', 'gutenberg_menu', 9 );
+
+if ( ! function_exists( 'html_contains_block' ) ) {
+	/**
+	 * Recursively find a block by attribute.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name The block type to search for.
+	 * @param  string  $attribute_name The attribute name to search for.
+	 * @param  string  $attribute_value The attribute value to search for.
+	 *
+	 * @return bool    True if block is found, false otherwise
+	 */
+	function html_contains_block( $html, $block_name, $attribute_name = null, $attribute_value = null ) {
+		$at = 0;
+
+		/**
+		 * This is the same regex as the one used in the block parser.
+		 * It is better to use this solution to look for a block's existence
+		 * in a document compared to having to parsing the blocks in the
+		 * document, avoiding all the performance drawbacks of achieving
+		 * a full representation of block content just to check if one block
+		 * is there.
+		 */
+		$pattern = sprintf(
+			'~<!--\s+?wp:%s\s+(?P<attrs>{(?:(?:[^}]+|}+(?=})|(?!}\s+/?-->).)*+)?}\s+)?/?-->~s',
+			// @TODO: we could ensure that there's a namespace on the block name.
+			preg_quote( $block_name, '~' )
+		);
+
+		while ( false !== preg_match( $pattern, $html, $matches, PREG_OFFSET_CAPTURE, $at ) ) {
+
+			// bail if no matches
```

Comment:
Not to be the bearer of bad news, but WordPress Coding Standards are going to throw a stink at comments like these. They need to start with a capital letter and end in a `.`